#!/usr/bin/bash
#run the whole recount unifier pipeline
#1) determine all finished (done) runs and symlink them in expected hierarchy
#2) run Snakemake summarization pipeline on output of 1)
#3) run QC stats on original file paths but requires the intron sums from 2)

#NOTE: script assumes:
#1) it's being run from a tranche/compilation specific subdirectory
#   under the recount-unifier repo checkout root directory
#2) ids.tsv is in the working directory, typically this is a link to the 
#   file generated by ../annotate/annotate_sjs.py that maps run accession to generated rail_id


#may need to update these with different paths/filenames/fileversions
annotated_sjs=../annotated_junctions.tsv.gz
existing_sums=../exons.bed.w_header.gz
gene_rejoin_mapping=../G029.G026.R109.F006.20190220.gtf.disjoint2exons2genes.bed
exon_rejoin_mapping=../G029.G026.R109.F006.20190220.gtf.disjoint2exons.bed
ref_sizes=../hg38.recount_pump.fa.new_sizes
ref_fasta=../hg38.recount_pump.fa

#compilation ID used in rail_id generation (e.g. tcga is 3, sra_human_v3_8 is 18, sra_human_v3_2 is 12)
comp_id=$1
#number of run accessions (SRRs)
num_runs=$2
threads=$3
#e.g. /storage2/cwilks/recount_pump/destination/srav3/sra_human_v3_9
original_full_path=$4
#e.g. tcga,sra_human_v3,etc...
study_prefix=$5

#example run for TCGA
#/bin/bash -x ../run_unifier.sh 3 11373 40 /storage2/cwilks/recount_pump/destination/tcga tcga
#or SRA human tranche 0:
#/bin/bash -x ../run_unifier.sh 10 30000 40 /storage2/cwilks/recount_pump/destination/srav3/sra_human_v3_0 sra_human_v3

ln -fs ../blank_exon_sums
ln -fs $original_full_path

/bin/bash -x ../scripts/find_done.sh $original_full_path links $study_prefix

timebin=`which time`

$timebin -v snakemake -j $threads --stats ./stats.json --snakefile ../Snakefile -p --config input=links staging=unified sample_ids_file=ids.tsv annotated_sjs=$annotated_sjs existing_sums=$existing_sums compilation_id=$comp_id gene_rejoin_mapping=$gene_rejoin_mapping exon_rejoin_mapping=$exon_rejoin_mapping num_samples=$num_runs ref_sizes=$ref_sizes ref_fasta=$ref_fasta > summarize.run 2>summarize.err

python3 ../log_qc/parse_logs_for_qc.py --incoming-dir links --sample-mapping ids.tsv --intron-sums intron_counts_summed.tsv > qc${comp_id}.tsv 2> qc${comp_id}.err
